man/vnamespace.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.41.2.
.TH VSPACE "8" "May 2013" "vspace  -- manages spaces" "System Administration"
.SH NAME
vspace \- manages spaces
.SH SYNOPSIS
.B vspace
\fI<operation> <spaces>* \fR[\fI--\fR] [\fI<program> <args>*\fR]
.SH DESCRIPTION
.SS "<operation> can be one of:"
.TP
\fB\-\-new\fR|\-n
\&...  create new spaces and execute <program> there;
<program> is mandatory in this case
.TP
\fB\-\-enter\fR|\-e <xid>
\&...  enter the spaces of context <xid> and execute
<program> there; <program> is mandatory in this
case
.TP
\fB\-\-set\fR|\-s
\&...  assign the current spaces to the current context
.TP
\fB\-\-index\fR|\-i <idx>
\&...  work with the namespaces for <idx>
.PP
<spaces>* specifies the spaces to manipulate.
It can be any combination of:
.TP
\fB\-\-mask\fR <mask>
\&...  specify a mask of spaces
.TP
\fB\-\-default\fR
\&...  the default spaces for this kernel
.TP
\fB\-\-mount\fR
\&...  the mount namespace
.TP
\fB\-\-fs\fR
\&...  the fs_struct
.TP
\fB\-\-ipc\fR
\&...  the IPC namespace
.TP
\fB\-\-uts\fR
\&...  the uts namespace
.TP
\fB\-\-user\fR
\&...  the user namespace
.TP
\fB\-\-pid\fR
\&...  the pid namespace
.TP
\fB\-\-net\fR
\&...  the network namespace
.PP
Please report bugs to vserver@list.linux\-vserver.org
.PP
This program is part of util\-vserver
.SH COPYRIGHT
Copyright \(co 2004 Enrico Scholz
.br
Copyright \(co 2007 Daniel Hokka Zakrisson
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.


NEWS
version 0.30.215
================

	- added support for using libnss instead of beecrypt for
	vhashify's hash calculation.  libnss has bad SHA1 performance on
	i386, but is a) maintained and b) gives better performance with
	all other hashes and on x86_64.

	- added tunctl, a small program to setup tun/tap interfaces, with
	support in the configuration.

	- vattribute and nattribute now support a --get operation, which
	parses the flags and capabilities into a human-readable form.

	- vspace supports all the spaces currently available.

	- vmemctrl controls the OOM bias of context.

	- dynamic contexts are unbroken.

	- hopefully all rpm-fake races have been fixed.

	- flock has been replaced by lockf, which works on NFS.

	- the clone build method now supports exclude lists.

	- vmount handles mounting into a guest.

	- various fixes for the Gentoo-scripts, thanks to Benedikt Böhm.

	- support for yum's persistdir has been added.

	- per-guest device maps are now supported.

	- vserver ... build now sets the barrier on the parent directory.

	- fsck is now supported for the guest's fstab.

	- symlinks are now handled on all levels in namespace cleanup.

	- the debootstrap build method now supports debootstrap 1.0.8.

	- a documentation update, thanks to Micah Anderson.

	- use numeric uids/gids for the template build method.

	- add support for the alpine and f8 distributions.
	
version 0.30.214
================
	- patches for yum-3.2.

	- various fixes to allow for *SuSE guests.

	- added Fedora 7.

	- replace file with h2ext for build -m template.

	- add exclude list support for vclone.

	- include debootstrap scripts for a lot of distributions.

	- fixes to allow the Gentoo initscripts to be used with baselayout 2.

	- remove disk limit caches on vserver ... delete.

	- added support for the new APIs in 2.2 and 2.3, in particular the
	netv2 API, allowing native IPv6 support.

	- added support for a separate filesystem tag.

version 0.30.213
================
	- added support for the device mapping feature present in development
	kernels.

	- further improved Gentoo support, again thanks to Benedikt Böhm.

	- support persistent contexts.

	- add another initscript that creates directories, configures vshelper
	as well as kills all contexts.

	- new vserver-stat implementation, which uses the accounting APIs
	available in Linux-VServer 2.2.

	- added support for setting sysctl values in the guest configuration.

	- added a clone build method, which simply links already unified files.

	- implement dynamic contexts in userspace, generating a static one
	during vserver ... build.

	- fix the endless loop in vlogin.

	- fix Ubuntu builds.

	- improved support for Debian Etch, thanks to Benedikt Böhm.

	- added CentOS 5.

	- use the state change helper to try to cleanup after us if the guest
	exited on its own.

version 0.30.212
================
	- support for the Linux-VServer 2.1 API has been implemented.

	- the scheduler configuration has been split up into multiple files.

	- namespace cleanup is enabled by default. This can be disabled by
	touch ${sysconfdir}/vservers/.defaults/nonamespace-cleanup

	- all unset ulimits are now reset to infinity, so limits are no longer
	inherited from the host.

	- much improved Gentoo support, thanks to Benedikt Böhm.

version 0.30.211
================

	- novlandev is now the default. To reenable the old behaviour of
	creating VLAN interfaces where used, touch
	${sysconfdir}/vservers/.defaults/interfaces/vlandev.

	- vshelper now works correctly with the kernel legacy support turned
	off.

	- vlogin provides a terminal proxy which allocates a new psuedo-tty
	on the inside when you enter a guest.

	- and lots of other features/fixes, see ChangeLog.


version 0.30.210
================

	- LEGACY BREAKAGE: by default, only recent kernel APIs will be
	supported so you have to make sure to use at least the vs2.0.1
	kernel patch. With older kernels, you have to call 'configure'
	with '--enable-apis=NOLEGACY'. Some programs will not work anymore
	with the default settings; these ones are:

	     + capchroot
	     * chcontext-context
	     * reducecap

	- dlimit support was integrated into the standard vserver
	operations. Thanks to Daniel Hokka Zakrisson for his patch.

	- recent kernel API is now used for setting up the network
	contexts. So it should be now possible to use a kernel compiled
	without CONFIG_VSERVER_LEGACYNET. Thanks to Daniel Hokka Zakrisson
	for his patch.  .


version 0.30.209
================

	- large revamping...
	
	
version 0.23.6
==============

	- fixed '--level' option on 'vserver ... chkconfig'


src/vnamespace.c
// $Id$    --*- c -*--

// Copyright (C) 2004 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include "util.h"
#include <lib_internal/sys_clone.h>

#include <vserver.h>

#include <getopt.h>
#include <libgen.h>
#include <errno.h>
#include <signal.h>
#include <sched.h>

#define ENSC_WRAPPERS_PREFIX	"vnamespace: "
#define ENSC_WRAPPERS_UNISTD	1
#define ENSC_WRAPPERS_VSERVER	1
#include <wrappers.h>

#define CMD_HELP		0x1000
#define CMD_VERSION		0x1001

int		wrapper_exit_code  =  255;

struct option const
CMDLINE_OPTIONS[] = {
  { "help",       no_argument,       0, CMD_HELP },
  { "version",    no_argument,       0, CMD_VERSION },
  { "new",        no_argument,       0, 'n' },
  { "enter",      required_argument, 0, 'e' },
  { "set",        no_argument,       0, 's' },
  { "cleanup",    no_argument,       0, 'c' },
  {0,0,0,0}
};

static void
showHelp(int fd, char const *cmd, int res)
{
  WRITE_MSG(fd, "Usage: ");
  WRITE_STR(fd, cmd);
  WRITE_MSG(fd,
	    " <operation> [--] [<program> <args>*]\n"
	    "\n"
	    "<operation> can be one of:\n"
	    "    --new|-n          ...  create new namespace and execute <program> there;\n"
	    "                           <program> is mandatory in this case\n"
	    "    --enter|-e <xid>  ...  enter the namespace of context <xid> and execute\n"
	    "                           <program> there; <program> is mandatory in this\n"
	    "                           case\n"
	    "    --set|-s          ...  make current namespace the namespace of the\n"
	    "                           current context\n"
	    "    --cleanup|-c      ...  remove all mounts from the namespace of the\n"
	    "                           current context\n"
	    "\n"
	    "Please report bugs to " PACKAGE_BUGREPORT "\n");

  exit(res);
}

static void
showVersion()
{
  WRITE_MSG(1,
	    "vnamespace " VERSION " -- manages filesystem-namespace\n"
	    "This program is part of " PACKAGE_STRING "\n\n"
	    "Copyright (C) 2004 Enrico Scholz\n"
	    VERSION_COPYRIGHT_DISCLAIMER);
  exit(0);
}

static void
newNamespace(void)
{
  pid_t		pid;

  signal(SIGCHLD, SIG_DFL);
  
#ifdef NDEBUG    
  pid = sys_clone(CLONE_NEWNS|CLONE_VFORK|SIGCHLD, 0);
#else
  pid = sys_clone(CLONE_NEWNS|SIGCHLD, 0);
#endif

  switch (pid) {
    case -1	:
      perror("vnamespace: clone()");
      exit(wrapper_exit_code);
    case 0	:
      break;
    default	:
      vc_exitLikeProcess(pid, wrapper_exit_code);
  }
}

static void
enterNamespace(xid_t xid, uint_least64_t mask)
{
  if (vc_enter_namespace(xid, mask)==-1) {
    perror("vnamespace: vc_enter_namespace()");
    exit(255);
  }
}

static void
setNamespace(xid_t xid, uint_least64_t mask)
{
  if (vc_set_namespace(xid, mask)==-1) {
    perror("vnamespace: vc_set_namespace()");
    exit(255);
  }
}

static void
cleanupNamespace()
{
  if (vc_cleanup_namespace()==-1) {
    perror("vnamespace: vc_cleanup_namespace()");
    exit(255);
  }
}

int main(int argc, char *argv[])
{
  bool		do_new     = false;
  bool		do_enter   = false;
  bool		do_set     = false;
  bool		do_cleanup = false;
  xid_t		xid        = VC_NOCTX;
  int		sum        = 0;
  
  while (1) {
    int		c = getopt_long(argc, argv, "+nsce:", CMDLINE_OPTIONS, 0);
    if (c==-1) break;

    switch (c) {
      case CMD_HELP	:  showHelp(1, argv[0], 0);
      case CMD_VERSION	:  showVersion();
      case 'n'		:  do_new     = true; break;
      case 's'		:  do_set     = true; break;
      case 'c'		:  do_cleanup = true; break;
      case 'e'		:
	do_enter = true;
	xid      = Evc_xidopt2xid(optarg,true);
	break;

      default		:
	WRITE_MSG(2, "Try '");
	WRITE_STR(2, argv[0]);
	WRITE_MSG(2, " --help' for more information.\n");
	return 255;
	break;
    }
  }

  sum = ((do_new ? 1 : 0) + (do_enter ? 1 : 0) +
	 (do_set ? 1 : 0) + (do_cleanup ? 1 : 0));
  
  if (sum==0)
    WRITE_MSG(2, "No operation was specified; try '--help' for more information\n");
  else if (sum>1)
    WRITE_MSG(2, "Can not specify multiple operations; try '--help' for more information\n");
  else if (optind==argc && (do_new || do_enter))
    WRITE_MSG(2, "No command specified; try '--help' for more information\n");
  else {
    if      (do_new)     newNamespace();
    else if (do_set)     setNamespace(VC_SAMECTX, CLONE_NEWNS|CLONE_FS);
    else if (do_cleanup) cleanupNamespace();
    else if (do_enter)   enterNamespace(xid, CLONE_NEWNS|CLONE_FS);

    if (optind<argc)
      EexecvpD(argv[optind], argv+optind);

    return EXIT_SUCCESS;
  }

  return 255;
}
